cloud/scope/powervs_cluster.go

```diff
@@ -157,3 +261,1549 @@ func (s *PowerVSClusterScope) PatchObject() error {
+		if err != nil {
+			if strings.Contains(err.Error(), "dhcp server does not exist") {
+				return nil
+			}
+			return fmt.Errorf("error fetching DHCP server: %w", err)
+		}
+
+		err = s.IBMPowerVSClient.DeleteDHCPServer(*server.ID)
+		if err != nil {
+			return fmt.Errorf("error deleting the DHCP server: %w", err)
+		}
+		s.Info("DHCP server successfully deleted")
+	}
+	return nil
+}
+
+// DeleteServiceInstance deletes service instance.
+func (s *PowerVSClusterScope) DeleteServiceInstance() error {
+	if !s.deleteResource(ServiceInstance) {
+		return nil
+	}
+
+	if s.IBMPowerVSCluster.Status.ServiceInstance.ID != nil {
+		serviceInstance, _, err := s.ResourceClient.GetResourceInstance(&resourcecontrollerv2.GetResourceInstanceOptions{
+			ID: s.IBMPowerVSCluster.Status.ServiceInstance.ID,
+		})
+		if err != nil {
+			return fmt.Errorf("error fetching service instance: %w", err)
+		}
+
+		if serviceInstance != nil && *serviceInstance.State == string(infrav1beta2.ServiceInstanceStateRemoved) {
+			return nil
+		}
+
+		servers, err := s.IBMPowerVSClient.GetAllDHCPServers()
+		if err != nil {
+			return fmt.Errorf("error fetching networks in the service instance: %w", err)
+		}
+
+		if len(servers) > 0 {
+			return fmt.Errorf("cannot delete service instance as network is not yet deleted")
+		}
+
+		_, err = s.ResourceClient.DeleteResourceInstance(&resourcecontrollerv2.DeleteResourceInstanceOptions{
+			ID: serviceInstance.ID,
+			//Recursive: pointer.Bool(true),
+		})
+
+		if err != nil {
+			s.Error(err, "error deleting Power VS service instance")
+			return err
+		}
+		s.Info("Service instance successfully deleted")
+	}
+	return nil
+}
+
+// DeleteCosInstance deletes COS instance.
+func (s *PowerVSClusterScope) DeleteCosInstance() error {
+	if !s.deleteResource(COSInstance) {
+		return nil
+	}
+
+	if s.IBMPowerVSCluster.Status.COSInstance.ID != nil {
+		cosInstance, _, err := s.ResourceClient.GetResourceInstance(&resourcecontrollerv2.GetResourceInstanceOptions{
+			ID: s.IBMPowerVSCluster.Status.COSInstance.ID,
+		})
+		if err != nil {
+			if strings.Contains(err.Error(), "COS instance unavailable") {
+				return nil
+			}
+			return fmt.Errorf("error fetching COS instance: %w", err)
+		}
+
+		if cosInstance != nil && (*cosInstance.State == "pending_reclamation" || *cosInstance.State == string(infrav1beta2.ServiceInstanceStateRemoved)) {
+			return nil
+		}
+
+		_, err = s.ResourceClient.DeleteResourceInstance(&resourcecontrollerv2.DeleteResourceInstanceOptions{
+			ID:        cosInstance.ID,
+			Recursive: pointer.Bool(true),
+		})
+		if err != nil {
+			s.Error(err, "error deleting COS service instance")
+			return err
+		}
+		s.Info("COS instance successfully deleted")
+	}
+	return nil
+}
+
+// deleteResource returns true or false to decide on deleting provided resource.
+func (s *PowerVSClusterScope) deleteResource(resourceType ResourceType) bool { //nolint:gocyclo
```

Comment:
or some other meaning func name. 
```suggestion
func (s *PowerVSClusterScope) resourceCreatedByController(resourceType ResourceType) bool { //nolint:gocyclo
```
same as other func, the name looks like, its deleting some resource